xml-parser/src/main/java/net/adoptopenjdk/icedteaweb/xmlparser/XMLParser.java

```diff
@@ -172,4 +178,218 @@ private static String getEncoding(final InputStream input) throws IOException {
+     * @param defaultValue default value
+     * @throws ParseException if the attribute does not exist or is empty
+     */
+    public static String getRequiredAttribute(final Node node, final String name, final String defaultValue, final boolean strict) throws ParseException {
+        final String result = getAttribute(node, name, null);
+
+        if (result == null || result.length() == 0) {
+            if (strict || defaultValue == null) {
+                throw new ParseException(R("PNeedsAttribute", node.getNodeName().getName(), name));
+            }
+        }
+
+        if (result == null) {
+            return defaultValue;
+        } else {
+            return result;
+        }
+    }
+
+    /**
+     * @return the same result as getURL except that a ParseException is thrown
+     * if the attribute is null or empty.
+     *
+     * @param node the node
+     * @param name the attribute containing an href
+     * @param base the base URL
+     * @throws ParseException if the JNLP file is invalid
+     */
+    public static URL getRequiredURL(final Node node, final String name, final URL base, final boolean strict) throws ParseException {
+        // probably should change "" to null so that url is always
+        // required even if !strict
+        getRequiredAttribute(node, name, "", strict);
+
+        return getURL(node, name, base, strict);
+    }
+
+    /**
+     * @return a URL object from a href string relative to the code base. If the
+     * href denotes a relative URL, it must reference a location that is a
+     * subdirectory of the codebase.
+     *
+     * @param node the node
+     * @param name the attribute containing an href
+     * @param base the base URL
+     * @throws ParseException if the JNLP file is invalid
+     */
+    public static URL getURL(final Node node, final String name, final URL base, final boolean strict) throws ParseException {
+        String href;
+        if (CODEBASE.equals(name)) {
+            href = getCleanAttribute(node, name);
+            //in case of null code can throw an exception later
+            //some bogus jnlps have codebase as "" and expect it behaving as "."
+            if (href != null && href.trim().isEmpty()) {
+                href = ".";
+            }
+        } else {
+            href = getAttribute(node, name, null);
+        }
+        return XMLParser.getURL(href, node.getNodeName().getName(), base, strict);
+    }
+
+    public static URL getURL(final String href, final String nodeName, final URL base, final boolean strict) throws ParseException {
+        if (href == null) {
+            return null; // so that code can throw an exception if attribute was required
+        }
+        try {
+            if (base == null) {
+                return new URL(href);
+            } else {
+                try {
+                    return new URL(href);
+                } catch (MalformedURLException ex) {
+                    // is relative
```

Comment:
In general I'm 100% with you. But this would be a point that changes functionality / the workflow. Until now I left Exception handling in all refactoring as it was. My plan was to do this later. What do you think?

Reply:
Yes we can postpone this